File: sample/iosApp/iosApp/MainViewModel.swift
//
//  MainViewModel.swift
//  iosApp
//
//  Created by Marco Gomiero on 19/03/23.
//  Copyright © 2023. All rights reserved.
//

import Foundation
import shared

class MainViewModel: ObservableObject {
    
    private let url = "https://www.androidauthority.com/feed"
    
    private let feedRepository = DI.shared.feedRepository
    
    @Published var uiState: UIState = UIState(
        isLoading: false,
        feed: nil,
        error: nil
    )
    
    func getFeed() {
        uiState = UIState(
            isLoading: true,
            feed: nil,
            error: nil
        )
        
        feedRepository.getFeed(url: url) { [self] feed, error in
            if error != nil {
                print("Something is wrong: \(String(describing: error))")
                DispatchQueue.main.async { [self] in
                    uiState = UIState(
                        isLoading: false,
                        feed: nil,
                        error: "Something went wrong"
                    )
                }
                
            }
            
            if let feed = feed {
                DispatchQueue.main.async { [self] in
                    uiState = UIState(
                        isLoading: false,
                        feed: feed,
                        error: nil
                    )
                }
            }
        }
    }
}

File: sample/androidApp/src/main/java/com/prof/rssparser/sample/android/MainViewModel.kt
package com.prof.rssparser.sample.android

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.State
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.prof.rssparser.sample.common.DI
import com.prof.rssparser.sample.common.UIState
import kotlinx.coroutines.launch

class MainViewModel : ViewModel() {

    private val url = "https://www.androidauthority.com/feed"

    private val feedRepository = DI.feedRepository

    private val _uiState: MutableState<UIState> = mutableStateOf(UIState())
    val uiState: State<UIState> = _uiState

    init {
        getFeeds()
    }

    fun getFeeds() {
        viewModelScope.launch {
            try {
                val feed = feedRepository.getFeed(url)
                _uiState.value = UIState(
                    isLoading = false,
                    feed = feed
                )
            } catch (e: Exception) {
                println("Something wrong while getting the feeds")
                e.printStackTrace()
                _uiState.value = UIState(
                    isLoading = false,
                    error = "Something wrong"
                )
            }
        }
    }
}
